cmd/calico-ipam-upgrade/main.go

```diff
@@ -0,0 +1,298 @@
+	calicoClient, err := client.New(*cfg)
+	if err != nil {
+		logCtxt.Fatal("failed to initialize api client")
+	}
+
+	// Execute migration
+	if err := Migrate(ctxt, calicoClient, nodename); err != nil {
+		logCtxt.WithError(err).Fatal("failed to migrate ipam")
+	}
+}
+
+func Migrate(ctxt context.Context, c client.Interface, nodename string) error {
+	// Set up logging formatting.
+	log.SetFormatter(&logutils.Formatter{})
+
+	// Install a hook that adds file/line no information.
+	log.AddHook(&logutils.ContextHook{})
+
+	// k8sClient directly calls the k8s apiserver.
+	k8sClient := c.(Accessor).Backend().(*k8s.KubeClient).ClientSet
+
+	var err error
+	defer func() {
+		if err != nil {
+			time.Sleep(10 * time.Second)
+		}
+	}()
+
+	// Establishing a mapping of IP addresses to Pods on this node.
+	// Check for IPIP tunnel address.
+	node, err := c.Nodes().Get(ctxt, nodename, options.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to get calico node resource: %s", err)
+	}
+
+	// Migrate IPIP tunnel address if the field is empty.
+	if node.Spec.BGP.IPv4IPIPTunnelAddr == "" {
+		log.Info("IPIP tunnel address not found, assigning...")
+
+		// Fetch k8s node.
+		k8sNode, err := k8sClient.CoreV1().Nodes().Get(nodename, metav1.GetOptions{})
+		if err != nil {
+			return fmt.Errorf("failed to get k8s node resource: %s", err)
+		}
+
+		// Parse PodCIDR.
+		ip, cidr, err := net.ParseCIDR(k8sNode.Spec.PodCIDR)
+		if err != nil {
+			return fmt.Errorf("PodCIDR %s did not parse successfully: %s", k8sNode.Spec.PodCIDR, err)
+		} else if cidr.Version() == 4 {
+			// We need to get the IP for the podCIDR and increment it to the
+			// first IP in the CIDR.
+			tunIp := ip.To4()
+			if tunIp == nil {
+				return fmt.Errorf("Cannot pick an IPv4 tunnel address from the given CIDR: %s", k8sNode.Spec.PodCIDR)
+			}
+			tunIp[3]++
+			node.Spec.BGP.IPv4IPIPTunnelAddr = tunIp.String()
+
+			// Assign the address via Calico IPAM.
+			ipipTunnelAddr := cnet.ParseIP(node.Spec.BGP.IPv4IPIPTunnelAddr)
+			if err = c.IPAM().AssignIP(ctxt, ipam.AssignIPArgs{
+				IP:       *ipipTunnelAddr,
+				Hostname: nodename,
+				Attrs: map[string]string{
+					"node":           nodename,
+					"ipipTunnelAddr": "true",
+				},
+			}); err != nil {
+				if _, ok := err.(errors.ErrorResourceAlreadyExists); !ok {
+					return fmt.Errorf("failed to get add IPIP tunnel addr %s: %s", node.Spec.BGP.IPv4IPIPTunnelAddr, err)
+				}
+				log.Info("IPIP tunnel address already assigned, continuing...")
+			}
+
+			// Save the calico node object to the datastore.
+			if node, err = c.Nodes().Update(ctxt, node, options.SetOptions{}); err != nil {
+				return fmt.Errorf("failed to save newly updated node object: %s", err)
+			}
+
+			log.Info("Successfully assigned and updated IPIP tunnel address")
+		}
+	}
+
+	// Check to see if the system is still using host-local
+	// by checking the existence of the path.
+	if _, err := os.Stat(ipAllocPath); err != nil && os.IsNotExist(err) {
```

Comment:
Shouldn't we do this first? We shouldn't try to migrate the IPIP tunnel address at all if there's no host-local directory IIUC.